tsdb/agent/db_test.go

```diff
@@ -934,6 +934,46 @@ func TestDBOutOfOrderTimeWindow(t *testing.T) {
 	}
 }
 
+func TestDBCreatedTimestampSamplesIngestion(t *testing.T) {
+	reg := prometheus.NewRegistry()
+	opts := DefaultOptions()
+	opts.OutOfOrderTimeWindow = 0
+	s := createTestAgentDB(t, reg, opts)
+	app := s.Appender(context.TODO())
+	defer s.Close()
+
+	lbls := labelsForTest(t.Name()+"_histogram", 1)
+	lset := labels.New(lbls[0]...)
+	_, err := app.AppendHistogramCTZeroSample(0, lset, 100, 30, tsdbutil.GenerateTestHistograms(1)[0], nil)
+	require.NoError(t, err)
+
+	_, err = app.AppendHistogramCTZeroSample(0, lset, 100, 100, tsdbutil.GenerateTestHistograms(1)[0], nil)
+	require.Error(t, err)
+
+	_, err = app.AppendHistogram(0, lset, 100, tsdbutil.GenerateTestHistograms(1)[0], nil)
+	require.NoError(t, err)
+	err = app.Commit()
+	require.NoError(t, err)
+
+	lbls = labelsForTest(t.Name(), 1)
+	lset = labels.New(lbls[0]...)
+	_, err = app.AppendCTZeroSample(0, lset, 100, 70)
+	require.NoError(t, err)
+
+	// not allow when CT is newer or equal to the sample timestamp
+	_, err = app.AppendCTZeroSample(0, lset, 100, 100)
+	require.Error(t, err)
+
+	_, err = app.Append(0, lset, 100, 0)
+	require.NoError(t, err)
+	err = app.Commit()
+	require.NoError(t, err)
+
+	m := gatherFamily(t, reg, "prometheus_agent_samples_appended_total")
+	require.Equal(t, float64(2), m.Metric[0].Counter.GetValue(), "agent wal mismatch of total appended samples")
+	require.Equal(t, float64(2), m.Metric[1].Counter.GetValue(), "agent wal mismatch of total appended histograms")
```

Comment:
urgh, it's so unfortunate that we need to check metrics to understand if things were appended or not :/ 

Checking metrics doesn't tell us what values/timestamps were appended, and therefore it doesn't catch the bugs that I've seen in the current code 😬

Reply:
I am now checking samples that were written to the WAL by the appender, I believe that is enough because once the writer gets notified it will read those same samples and just remote write.